src/fields/fft_poisson_solver/fft/WrapCuDST.cpp

```diff
@@ -67,81 +69,316 @@ namespace AnyDST
+    {
+        HIPACE_PROFILE("AnyDST::C2Rfft()");
+        cudaStream_t stream = amrex::Gpu::Device::cudaStream();
+        cufftSetStream(plan, stream);
+        cufftResult result;
+
+#ifdef AMREX_USE_FLOAT
+        result = cufftExecC2R(plan, reinterpret_cast<AnyFFT::Complex*>(in), out);
+#else
+        result = cufftExecZ2D(plan, reinterpret_cast<AnyFFT::Complex*>(in), out);
+#endif
+        if ( result != CUFFT_SUCCESS ) {
+            amrex::Print() << " forward transform using cufftExec failed ! Error: " <<
+                CuFFTUtils::cufftErrorToString(result) << "\n";
+        }
+    };
+
+    /** \brief Make Sine-space Real array out of array from fft.
+     * out[idx] = 0.5 *(in[n_data-idx] - in[idx+1] + (in[n_data-idx] + in[idx+1])/
+     * (2*sin((idx+1)*pi/(n_data+1)))) for each column
+     *
+     * \param[in] in input real array
+     * \param[out] out output real array
+     * \param[in] n_data number of (contiguous) rows in position matrix
+     * \param[in] n_batch number of (strided) columns in position matrix
+     */
+    void ToSine (const amrex::Real* const AMREX_RESTRICT in, amrex::Real* const AMREX_RESTRICT out,
+                 const int n_data, const int n_batch)
+    {
+        HIPACE_PROFILE("AnyDST::ToSine()");
+        using namespace amrex::literals;
+
+        const amrex::Real n_1_real = n_data+1._rt;
+        const int n_1 = n_data+1;
+        amrex::ParallelFor({{1,0,0}, {(n_data+1)/2,n_batch-1,0}},
+            [=] AMREX_GPU_DEVICE(int i, int j, int) noexcept
+            {
+                const int i_rev = n_1-i;
+                const int stride_in = n_1*j;
+                const int stride_out = n_data*j;
+                const amrex::Real in_a = in[i+stride_in];
+                const amrex::Real in_b = in[i_rev+stride_in];
+#ifdef AMREX_USE_FLOAT
+                out[i-1+stride_out] = 0.5_rt*(in_b-in_a+(in_a+in_b)/(2._rt*sinpif(i/n_1_real)));
+                out[i_rev-1+stride_out] = 0.5_rt*(in_a-in_b+(in_a+in_b)/(2._rt
+                                          *sinpif(i_rev/n_1_real)));
+#else
+                out[i-1+stride_out] = 0.5_rt*(in_b-in_a+(in_a+in_b)/(2._rt*sinpi(i/n_1_real)));
+                out[i_rev-1+stride_out] = 0.5_rt*(in_a-in_b+(in_a+in_b)/(2._rt
+                                          *sinpi(i_rev/n_1_real)));
+#endif
+            });
+    };
+
+    /** \brief Transpose input matrix
+     * out[idy][idx] = in[idx][idy]
+     *
+     * \param[in] in input real array
+     * \param[out] out output real array
+     * \param[in] n_data number of (contiguous) rows in input matrix
+     * \param[in] n_batch number of (strided) columns in input matrix
+     */
+    void Transpose (const amrex::Real* const AMREX_RESTRICT in,
+                    amrex::Real* const AMREX_RESTRICT out,
+                    const int n_data, const int n_batch)
+    {
+        HIPACE_PROFILE("AnyDST::Transpose()");
+        constexpr int tile_dim = 32;
```

Comment:
Just a suggestion to make it a bit safer, in case we need to play with this parameter in the future.
```suggestion
        constexpr int tile_dim_power = 5;
        const int tile_dim = std::pow(2, tile_dim_power);
```
You would probably need to add
```c++
#include <cmath>
```
at the end of the includes at the beginning of this file.

Reply:
Looking at Compiler Explorer, I found that with `constexpr int i = 32;` `num>>5` and `num/i` are exactly equivalent for gcc, clang and nvcc, as long as `num` is an `unsigned int`.
With default settings, gcc was also able to do the same thing with `const int i = std::pow(2,5);`, but not clang which would actually call std::pow.
I think it’s best to replace the `>>5` with a divide and add `unsigned` wherever possible.

Reply:
Agreed